.vscode/settings.json

```
@@ -3,6 +3,6 @@
   "editor.formatOnSave": true,
   "editor.formatOnPaste": false,
   "[python]": {
-    "editor.defaultFormatter": "ms-python.black-formatter"
+    "editor.defaultFormatter": "charliermarsh.ruff"
   }
```

I tried this out locally with VSCode. Can we try these two suggestions -- use the bundled version of ruff instead of looking for one in the Python virtualenv (since we don't install it by default), and only run on save (default is to run on every keystroke).

```suggestion
  },
  "ruff.importStrategy": "useBundled",
  "ruff.lint.run": "onSave"
```

With more usage we may find other tweaks but this is the minimum I needed to do to make this work as expected in VSCode for me. Thanks for considering this.